android/KMEA/app/src/test/java/com/tavultesoft/kmea/logic/ResourceUpdateToolTest.java

```diff
@@ -0,0 +1,72 @@
+package com.tavultesoft.kmea.logic;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+import androidx.test.core.app.ApplicationProvider;
+
+import com.tavultesoft.kmea.R;
+
+import org.junit.Assert;
+import org.junit.Assume;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+
+@RunWith(RobolectricTestRunner.class)
+public class ResourceUpdateToolTest {
+
+  @Test
+  public void shouldUpdateFirstCallTest() {
+    ResourcesUpdateTool _updateTool = new ResourcesUpdateTool();
+    Assert.assertTrue(_updateTool.shouldCheckUpdate(ApplicationProvider.getApplicationContext()));
+  }
+
+  @Test
+  public void shouldUpdateFalseCallTest() {
+    Assume.assumeFalse(ResourcesUpdateTool.FORCE_RESOURCE_UPDATE);
+
+    Calendar _cal = GregorianCalendar.getInstance();
+    _cal.add(Calendar.HOUR_OF_DAY,-5);
+
+    setLastUpdatePref(_cal);
+
+    ResourcesUpdateTool _updateTool = new ResourcesUpdateTool();
+    Assert.assertFalse(_updateTool.shouldCheckUpdate(ApplicationProvider.getApplicationContext()));
+  }
+
+  @Test
+  public void shouldUpdateFalseCallTest2() {
+    Assume.assumeFalse(ResourcesUpdateTool.FORCE_RESOURCE_UPDATE);
+
+    Calendar _cal = GregorianCalendar.getInstance();
+
+    setLastUpdatePref(_cal);
+
+    ResourcesUpdateTool _updateTool = new ResourcesUpdateTool();
+    Assert.assertFalse(_updateTool.shouldCheckUpdate(ApplicationProvider.getApplicationContext()));
+  }
+
+  @Test
+  public void shouldUpdateTrueCallTest() {
+    Calendar _cal = GregorianCalendar.getInstance();
+    _cal.add(Calendar.DAY_OF_MONTH,-1);
+    _cal.add(Calendar.MINUTE,-1);
+    setLastUpdatePref(_cal);
+
+    ResourcesUpdateTool _updateTool = new ResourcesUpdateTool();
+    Assert.assertTrue(_updateTool.shouldCheckUpdate(ApplicationProvider.getApplicationContext()));
+  }
+
+  private void setLastUpdatePref(Calendar theLastUpdate) {
+    SharedPreferences prefs = ApplicationProvider.getApplicationContext().getSharedPreferences(
+      ApplicationProvider.getApplicationContext().getString(R.string.kma_prefs_name), Context.MODE_PRIVATE);
+
+    SharedPreferences.Editor editor = prefs.edit();
+    editor.putLong("lastUpdateCheck", theLastUpdate.getTimeInMillis());
```

Comment:
1. Is the shared preference "lastUpdateCheck" only used for this unit test? If the app also uses the preference, maybe add a teardown step to remove the preference so a clean install of the app won't have it set.

2. Maybe define `lastUpdateCheck` as a constant key.

Reply:
I think the tests have a new clean Context on every startup.
They should not have any effects on your virtual devices